### DISCUSSION TITLE
Varying bandwidth and/or lengths

### DISCUSSION THREAD

**Discussion by swg0101** [esnet/iperf#1322] [Ideas]
Currently, I see that iperf3 uses the same packet size for all packets - this generally works in most cases, but is there any option I can use that can vary the packet lengths that are being sent by a specific endpoint (e.g. 220-238 bytes) so that it mimics more or less what real-life traffic may look like for a specific application?

Also, it seems like the same value is currently used under bidirectional mode. Many times this is not feasible because the other side may not have a link that allows symmetric throughput (hence forcing you to use the lowest common denominator that can be used for both parties). As an example, my cable link supports up to 600Mbps download, but the uplink is only 20Mbps. If I were to use the bidirectional test, I would not be able to test above 20Mbps without saturating the uplink. It would be great if one can specify the value for the other side separately, e.g. (use 20Mbps for downstream but 2 Mbps for upstream - best if we can use a combination of various lengths and bandwidths for both sides).

**Comment by davidBar-On** [comment]
I have submitted two PRs that may help:

1. PR #1093 allows specifying different bandwidth limits for the two bidir directions (Hopefully this PR will be merged into master sometime.)
2. PR #1075 allows sending random UDP block sizes and adding random delays between the packets sent (since TCP is a steam protocol, there is not much meaning to such parameters for TCP).  The delays allow to better simulate real traffic as standard iperf3 behavior is to send all data per interval at the beginning of the interval, which is good for throughput test but not simulates real traffic.  (Note that this is simplified version of PR #1004 that also supported fixed intervals between block sizes (instead of random) and the ability to send bursts of packets.)
